src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,169 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    private static class Node<T> {
+        private T element;
+        private Node<T> next;
+        private Node<T> previous;
+
+        public Node(T element) {
+            this.element = element;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (first == null) {
+            // Список порожній
+            first = newNode;
+            last = newNode;
+        } else {
+            last.next = newNode;
+            newNode.previous = last;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bound ");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> newNode = new Node<>(value);
+        if (index == 0) {
+            newNode.next = first;
+            first.previous = newNode;
+            first = newNode;
+        } else {
+            Node<T> indexNode = first;
+            for (int i = 0; i < index; i++) {
+                indexNode = indexNode.next;
+            }
+            newNode.previous = indexNode.previous;
+            newNode.next = indexNode;
+            indexNode.previous.next = newNode;
+            indexNode.previous = newNode;
+        }
+        size++;
     }
```

Comment:
Consider creating a `findNodeByIndex` method to avoid code duplication and a `checkIndex` method to verify the index is valid. Also, don't iterate your nodes from the head to the tail; depending on the index value, you can iterate only the first or second half of the list to boost performance.